src/docfx/build/moniker/MonikersProvider.cs

```diff
@@ -3,23 +3,44 @@
 
 using System;
 using System.Collections.Generic;
+using System.IO;
 using System.Linq;
 
 namespace Microsoft.Docs.Build
 {
     internal class MonikersProvider
     {
         private readonly List<(Func<string, bool> glob, string monikerRange)> _rules = new List<(Func<string, bool>, string)>();
+        private readonly MonikerRangeParser _rangeParser;
+        private readonly Lazy<MonikerComparer> _monikerAscendingComparer;
+        private readonly Lazy<MonikerComparer> _monikerDescendingComparer;
 
-        public MonikersProvider(Config config)
+        public MonikerComparer AscendingComparer => _monikerAscendingComparer.Value;
+
+        public MonikerComparer DescendingComparer => _monikerDescendingComparer.Value;
+
+        public MonikersProvider(Docset docset, Config config)
         {
             foreach (var (key, monikerRange) in config.MonikerRange)
             {
                 _rules.Add((GlobUtility.CreateGlobMatcher(key), monikerRange));
             }
             _rules.Reverse();
+
+            var monikerDefinition = new MonikerDefinitionModel();
+            if (!string.IsNullOrEmpty(config.MonikerDefinition))
+            {
+                var path = docset.GetFileRestorePath(config.MonikerDefinition);
+                monikerDefinition = JsonUtility.Deserialize<MonikerDefinitionModel>(File.ReadAllText(path));
+            }
+            _rangeParser = new MonikerRangeParser(monikerDefinition);
+            _monikerAscendingComparer = new Lazy<MonikerComparer>(() => new MonikerComparer(monikerDefinition));
```

Comment:
`Lazy<>` is not needed here, right? just initiate a new comparer

Reply:
Also, we have `Lazy<MonikersProvider>` in the outer layer

Reply:
ok


---
In reply to: [237374624](https://github.com/dotnet/docfx/pull/3695#discussion_r237374624) [](ancestors = 237374624)